File: plugin-dotnet-agent/src/main/java/jetbrains/buildServer/dotnet/DotnetToolProvider.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.dotnet;

import jetbrains.buildServer.agent.*;
import jetbrains.buildServer.util.StringUtil;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Lookups for .NET CLI utilities.
 */
public class DotnetToolProvider implements ToolProvider {

    private static final Pattern TOOL_PATTERN = Pattern.compile("^.*" + DotnetConstants.RUNNER_TYPE + "(\\.(exe))?$");
    private static final String PATH_VARIABLE = "PATH";
    private static final String UNABLE_TO_LOCATE_TOOL = "Unable to locate tool %s in the system. " +
            "Please make sure that `%s` variable contains .NET CLI toolchain directory or " +
            "defined `%s` variable.";

    public DotnetToolProvider(@NotNull final ToolProvidersRegistry toolProvidersRegistry) {
        toolProvidersRegistry.registerToolProvider(this);
    }

    @Override
    public boolean supports(@NotNull String toolName) {
        return DotnetConstants.RUNNER_TYPE.equalsIgnoreCase(toolName);
    }

    @NotNull
    @Override
    public String getPath(@NotNull String toolName) throws ToolCannotBeFoundException {
        final String pathVariable = System.getenv(PATH_VARIABLE);
        final List<String> paths = StringUtil.splitHonorQuotes(pathVariable, File.pathSeparatorChar);

        // Try to use DOTNET_HOME variable
        final String dotnetHomeVariable = System.getenv(DotnetConstants.TOOL_HOME);
        if (!StringUtil.isEmpty(dotnetHomeVariable)) {
            paths.add(0, dotnetHomeVariable);
        }

        final String toolPath = FileUtils.findToolPath(paths, TOOL_PATTERN);
        if (StringUtil.isEmpty(toolPath)) {
            final String message = String.format(UNABLE_TO_LOCATE_TOOL, toolName, PATH_VARIABLE, DotnetConstants.TOOL_HOME);
            throw new ToolCannotBeFoundException(message);
        }

        return toolPath;
    }

    @NotNull
    @Override
    public String getPath(@NotNull String toolName,
                          @NotNull AgentRunningBuild build,
                          @NotNull BuildRunnerContext runner) throws ToolCannotBeFoundException {
        return getPath(toolName);
    }
}

File: plugin-dotnet-agent/src/main/kotlin/jetbrains/buildServer/dotnet/DotnetToolProvider.kt


package jetbrains.buildServer.dotnet

import jetbrains.buildServer.agent.*
import jetbrains.buildServer.agent.Logger
import jetbrains.buildServer.dotnet.discovery.dotnetSdk.DotnetSdksProvider
import java.io.File

/**
 * Lookups for .NET CLI utilities.
 */
class DotnetToolProvider(
        toolProvidersRegistry: ToolProvidersRegistry,
        private val _toolSearchService: ToolSearchService,
        private val _toolEnvironment: ToolEnvironment,
        private val _dotnetSdksProviderImpl: DotnetSdksProvider
)
    : ToolProvider {
    init {
        toolProvidersRegistry.registerToolProvider(this)
    }

    override fun supports(toolName: String): Boolean = DotnetConstants.RUNNER_TYPE.equals(toolName, ignoreCase = true)

    override fun getPath(toolName: String): String =
            if(supports(toolName))
                executablePath
                ?.absolutePath
                ?: throw ToolCannotBeFoundException("""
                        Unable to locate tool $toolName in the system. Please make sure that `PATH` variable contains
                        .NET CLI toolchain directory or defined `${DotnetConstants.TOOL_HOME}` variable.""".trimIndent())
            else
                throw ToolCannotBeFoundException("Unsupported tool.")

    private val executablePath: File? by lazy {
        val executables = _toolSearchService.find(DotnetConstants.EXECUTABLE, _toolEnvironment.homePaths + _toolEnvironment.defaultPaths + _toolEnvironment.environmentPaths).toList()
        for (dotnetExecutable in executables) {
            if (_dotnetSdksProviderImpl.getSdks(dotnetExecutable).any()) {
                return@lazy dotnetExecutable
            }
            else {
                LOG.debug("Cannot find .NET Core SDK for <${dotnetExecutable}>.")
            }
        }

        return@lazy executables.firstOrNull()
    }

    @Throws(ToolCannotBeFoundException::class)
    override fun getPath(toolName: String, build: AgentRunningBuild, runner: BuildRunnerContext): String {
        return if (runner.virtualContext.isVirtual) {
            DotnetConstants.EXECUTABLE
        } else {
            getPath(toolName)
        }
    }

    companion object {
        private val LOG = Logger.getLogger(DotnetToolProvider::class.java)
    }
}